awx_collection/plugins/module_utils/tower_connector.py

```diff
@@ -0,0 +1,446 @@
+            data = dumps(kwargs.get('data', {}))
+
+        try:
+            response = self.session.open(method, self.url.geturl(), headers=headers, validate_certs=self.verify_ssl, follow_redirects=True, data=data)
+        except(SSLValidationError) as ssl_err:
+            raise AnsibleTowerException("Could not establish a secure connection to your host ({1}): {0}.".format(self.url.netloc, ssl_err))
+        except(ConnectionError) as con_err:
+            raise AnsibleTowerException("There was a network error of some kind trying to connect to your host ({1}): {0}.".format(self.url.netloc, con_err))
+        except(HTTPError) as he:
+            # Sanity check: Did the server send back some kind of internal error?
+            if he.code >= 500:
+                raise AnsibleTowerException('The host sent back a server error ({1}): {0}. Please check the logs and try again later'.format(self.url.path, he))
+            # Sanity check: Did we fail to authenticate properly?  If so, fail out now; this is always a failure.
+            elif he.code == 401:
+                raise AnsibleTowerException('Invalid Tower authentication credentials for {0} (HTTP 401).'.format(self.url.path))
+            # Sanity check: Did we get a forbidden response, which means that the user isn't allowed to do this? Report that.
+            elif he.code == 403:
+                raise AnsibleTowerException("You don't have permission to {1} to {0} (HTTP 403).".format(self.url.path, method))
+            # Sanity check: Did we get a 404 response?
+            # Requests with primary keys will return a 404 if there is no response, and we want to consistently trap these.
+            elif he.code == 404:
+                if kwargs.get('return_none_on_404', False):
+                    return None
+                raise AnsibleTowerException('The requested object could not be found at {0}.'.format(self.url.path))
+            # Sanity check: Did we get a 405 response?
+            # A 405 means we used a method that isn't allowed. Usually this is a bad request, but it requires special treatment because the
+            # API sends it as a logic error in a few situations (e.g. trying to cancel a job that isn't running).
+            elif he.code == 405:
+                raise AnsibleTowerException("The Tower server says you can't make a request with the {0} method to this endpoing {1}".format(method, self.url.path))
+            # Sanity check: Did we get some other kind of error?  If so, write an appropriate error message.
+            elif he.code >= 400:
+                # We are going to return a 400 so the module can decide what to do with it
+                page_data = he.read()
+                try:
+                    return {'status_code': he.code, 'json': loads(page_data)}
+                # JSONDecodeError only available on Python 3.5+
+                except ValueError:
+                    return {'status_code': he.code, 'text': page_data}
```

Comment:
I have never liked this exception for text, and it becomes much more painful when part of this connector class.

In what actual scenario did the server ever give non-JSON data for a 400 response? If you think about how a module should behave in that scenario, does this behavior of the connector class still make sense? Probably not.

One possible case is that the proper headers were not passed, so you get an HTML response. That should not be a case where a subtly different dict is returned, that should be a textbook `AnsibleTowerException`. It's never okay to get HTML data. Sure, there was a 400, there are multiple problems, and these could all be provided in different keys to present to the user in `fail_json`. Handling the 400 error should not be the responsibility of the module in that case. Actually, it should be very rare that handling that is left to the module.

Reply:
I updated this second return to raise an AnsibleTowerException rather than do a return.